keras/src/ops/core.py

```diff
@@ -199,6 +201,99 @@ def scan(f, init, xs, length=None):
     )
 
 
+class AssociativeScan(Operation):
+    def __init__(self, reverse=False):
+        super().__init__()
+        self.reverse = reverse
+
+    def call(self, f, elems, axis=0):
+        return backend.core.associative_scan(
+            f, elems, reverse=self.reverse, axis=axis
+        )
+
+    def compute_output_spec(self, f, elems, axis):
+        elems_flat, tree_ = optree.tree_flatten(elems)
+        lens = [elem.shape[axis] for elem in elems_flat]
+        if len(set(lens)) != 1:
+            raise ValueError(
+                "Array inputs to associative_scan must have the same "
+                "first dimension. (saw: {})".format(
+                    [elem.shape for elem in elems_flat]
+                )
+            )
+
+        x = optree.tree_unflatten(
+            tree_, [slice_along_axis(x, 0, 1, axis=axis) for x in elems_flat]
+        )
+        y_spec = backend.compute_output_spec(f, x, x)
+
+        def _restore_shape(x):
+            return KerasTensor(
+                shape=elems_flat[0].shape, dtype=x.dtype, sparse=x.sparse
+            )
+
+        y_spec = tree.map_structure(_restore_shape, y_spec)
+        return y_spec
+
+
+@keras_export("keras.ops.associative_scan")
+def associative_scan(f, elems, reverse=False, axis=0):
+    """Performs a scan with an associative binary operation, in parallel.
+
+    This operation has a similar use-case to scan. The key difference is that
```

Comment:
Thanks for the detailed feedback! I'd also like to add a little bit of context to this statement using some ad hoc benchmarking results from [this script](https://gist.github.com/SeKim12/0b5a77fbb05c707e60dcee03cfd7c24b). I have results for different backends as well showing a similar trend. 

![numpy_result](https://github.com/keras-team/keras/assets/63029244/d5e0e743-27fb-4700-8170-2a5e4320c8dc)